apps/pollbook/backend/src/networking.ts

```diff
@@ -261,41 +287,80 @@ export function setupMachineNetworking({
           debug(
             'The current service is no longer found on avahi. Setting online status to false'
           );
-          workspace.store.setOnlineStatus(false);
+          workspace.store.setPollbookServiceForName(selfMachineServiceName, {
+            ...selfMachinePollbookService,
+            apiClient: undefined,
+            address: undefined,
+            status: PollbookConnectionStatus.LostConnection,
+          });
           return;
         }
         for (const { name, resolvedIp, port } of services) {
           debug('Checking service %s at %s:%d', name, resolvedIp, port);
           if (
-            name !== currentNodeServiceName &&
+            name !== selfMachineServiceName &&
             !workspace.store.getIsOnline()
           ) {
             // do not bother trying to ping other nodes if we are not online
             continue;
           }
-          const currentPollbookService = previouslyConnected[name];
-          if (currentPollbookService && currentPollbookService.apiClient) {
-            debug('Using previous API client for %s', name);
+          const currentPollbookService = previousPollbookServices[name];
+
+          const advertisedServiceUrl = `http://${resolvedIp}:${port}`;
+          // We want to re-use the old api client if it is set up, and the URL it is using matches the advertised URL.
+          // Avahi can in rare cases return an invalid ipv6 address in the ipv4 field. To mitigate the impact of that when we
+          // see an invalid address we will continue to use the last known good address and api client (if it exists).
+          const reuseApiClient =
+            // We have a valid old api client
+            currentPollbookService &&
+            currentPollbookService.apiClient !== undefined &&
+            // It is for the same address we just got, or the new one is invalid
+            (currentPollbookService.address === advertisedServiceUrl ||
+              !isValidIpv4Address(resolvedIp));
+          if (reuseApiClient) {
+            debug(
+              'Using previous API client for %s with url %s',
+              name,
+              currentPollbookService.address
+            );
+          } else {
+            debug(
+              'Creating new API client for %s with url %s',
+              name,
+              advertisedServiceUrl
+            );
           }
-          const apiClient =
-            currentPollbookService && currentPollbookService.apiClient
-              ? currentPollbookService.apiClient
-              : createPeerApiClientForAddress(`http://${resolvedIp}:${port}`);
+          const machineUrl = reuseApiClient
+            ? assertDefined(currentPollbookService.address)
+            : advertisedServiceUrl;
+
+          const apiClient = reuseApiClient
+            ? assertDefined(currentPollbookService.apiClient)
+            : createPeerApiClientForAddress(machineUrl);
```

Comment:
wondering if it might be slightly easier to follow if we just structure this as `if (reuseApiClient) { ... } else { ... }` instead of inlining ternary operators multiple times. just my personal preference, not blocking.

Reply:
I find ternary operators better when setting variables though so you can use const